### ISSUE TITLE
torch.cumsum at last dimension

### ISSUE THREAD

**Issue by 7starsea** [pytorch/pytorch#59860]
## 🐛 Bug

I am expecting that the results of `torch.cumsum` of the same tensor should be exactly the same for different device (here cpu, cuda).
If the parameter `dim` is not the last dimension, then everything is fine. However, if the parameter `dim` is the last dimension, there is minor difference between the results in cpu and cuda. I know most of the times the minor difference is acceptable, but sometimes, the minor difference will incur significant results in cpu and cuda. 

## To Reproduce
```
import torch
x = torch.rand(4000, 20, 10, 20, dtype=torch.float64)
x_cu = x.cuda()

for dim in range(x.ndim):
    y = torch.cumsum(x, dim=dim)
    y_cu = torch.cumsum(x_cu, dim=dim)
    err = torch.sum(torch.abs(y_cu.cpu() - y))
    print(dim, err)
```

## Expected behavior
```
0 tensor(0., dtype=torch.float64)
1 tensor(0., dtype=torch.float64)
2 tensor(0., dtype=torch.float64)
3 tensor(0., dtype=torch.float64)
```
but usually, for the last dimension, we could only have
```
3 tensor(6.6426e-09, dtype=torch.float64)
```

## Environment

PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.18.2
Libc version: glibc-2.27

Python version: 3.8 (64-bit runtime)
Python platform: Linux-4.15.0-143-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 9.1.85
GPU models and configuration: 
GPU 0: TITAN V
GPU 1: TITAN V

Nvidia driver version: 450.119.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1
[pip3] torchaudio==0.8.0a0+e4e171a
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               10.2.89              hfd86e86_1    defaults
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.2                      256    defaults
[conda] mkl-service               2.3.0            py38he904b0f_0    defaults
[conda] mkl_fft                   1.2.0            py38h23d657b_0    defaults
[conda] mkl_random                1.1.1            py38h0573a6f_0    defaults
[conda] numpy                     1.19.2           py38h54aff64_0    defaults
[conda] numpy-base                1.19.2           py38hfa32c7d_0    defaults
[conda] pytorch                   1.8.1           py3.8_cuda10.2_cudnn7.6.5_0    pytorch
[conda] torchaudio                0.8.1                      py38    pytorch
[conda] torchvision               0.9.1                py38_cu102    pytorch

**Comment by ngimel** [comment]
This is expected, cuda and cpu are using different algorithms to compute cumsum.

**Comment by 7starsea** [comment]
> This is expected, cuda and cpu are using different algorithms to compute cumsum.

I am curious why the minor difference only occurs at the last dimension?
